### PR TITLE
Handle backup secret gossip

### PR DESCRIPTION
**Pull request by BillCarsonFr**
Fixes https://github.com/vector-im/element-web/issues/26288
Second step will be to automatically restore after the secret is received (in a following PR)

## Checklist

-   [x] Tests written for new code (and old code if feasible)
-   [x] Linter and other CI checks pass
-   [x] Sign-off given on the changes (see [CONTRIBUTING.md](https://github.com/matrix-org/matrix-js-sdk/blob/develop/CONTRIBUTING.md))

---
This change is marked as an *internal change* (Task), so will not be included in the changelog.

### CODE REVIEW

File: spec/integ/crypto/verification.spec.ts [hunk 1]
```diff
@@ -1160,6 +1172,296 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("verification (%s)", (backend: st
...
+
+            // The client should now know about the olm device
+            const devices = await aliceClient.getCrypto()!.getUserDeviceInfo([TEST_USER_ID]);
+            expect(devices.get(TEST_USER_ID)!.keys()).toContain(olmDeviceId);
+        });
+
+        afterEach(async () => {
+            if (aliceClient !== undefined) {
+                aliceClient.stopClient();
+            }
```
Comment (florianduros) [inline spec/integ/crypto/verification.spec.ts:1237]:
```suggestion
            aliceClient?.stopClient();

```

File: spec/integ/crypto/verification.spec.ts [hunk 2]
```diff
@@ -1160,6 +1172,296 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("verification (%s)", (backend: st
...
+
+            if (testOlmAccount) {
+                testOlmAccount.free();
+            }
```
Comment (florianduros) [inline spec/integ/crypto/verification.spec.ts:1241]:
```suggestion
            testOlmAccount?.free();
```

File: spec/integ/crypto/verification.spec.ts [hunk 3]
```diff
@@ -1160,6 +1172,296 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("verification (%s)", (backend: st
...
+            const requestId = await requestPromises.get("m.megolm_backup.v1");
+
+            await sendBackupGossipAndExpectVersion(requestId!, BACKUP_DECRYPTION_KEY_BASE64, matchingBackupInfo);
+
+            // We are lacking a way to signal that the secret has been received, so we wait a bit..
+            jest.useRealTimers();
+            await new Promise((resolve) => {
+                setTimeout(resolve, 500);
+            });
+            jest.useFakeTimers();
```
Comment (florianduros) [inline spec/integ/crypto/verification.spec.ts:1273]: Can't we use `flushPromises`or something with the fake timers ?

File: spec/integ/crypto/verification.spec.ts [hunk 4]
```diff
@@ -1160,6 +1172,296 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("verification (%s)", (backend: st
...
+        newBackendOnly("Should not accept the backup decryption key gossip if private key do not match", async () => {
+            const requestPromises = mockUpSecretRequestAndGetPromises();
+
+            await doInteractiveVerification();
+
+            const requestId = await requestPromises.get("m.megolm_backup.v1");
+
+            await sendBackupGossipAndExpectVersion(requestId!, BACKUP_DECRYPTION_KEY_BASE64, nonMatchingBackupInfo);
+
+            // We are lacking a way to signal that the secret has been received, so we wait a bit..
```
Comment (florianduros) [inline spec/integ/crypto/verification.spec.ts:1290]: If we need to this operation in multiple tests, we can extract it in a dedicated function in this test suite

File: src/rust-crypto/backup.ts [hunk 5]
```diff
@@ -106,6 +106,31 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
...
+        try {
+            backupDecryptionKey = RustSdkCryptoJs.BackupDecryptionKey.fromBase64(privateKeyBase64);
+        } catch (e) {
+            logger.warn("backupMatchesPrivateKey: Invalid backup decryption key", e);
+            return false;
+        }
+
+        return (
+            (info.auth_data as Curve25519AuthData)?.public_key === backupDecryptionKey.megolmV1PublicKey.publicKeyBase64
+        );
```
Comment (florianduros) [inline src/rust-crypto/backup.ts:131]:
```suggestion
        try {
            const backupDecryptionKey = RustSdkCryptoJs.BackupDecryptionKey.fromBase64(privateKeyBase64);
             return (info.auth_data as Curve25519AuthData)?.public_key === backupDecryptionKey.megolmV1PublicKey.publicKeyBase64;
        } catch (e) {
            logger.warn("backupMatchesPrivateKey: Invalid backup decryption key", e);
            return false;
        }
```

**Review by florianduros — CHANGES_REQUESTED** [review]

File: spec/integ/crypto/olm-utils.ts [hunk 6]
```diff
@@ -218,6 +330,44 @@ export function encryptGroupSessionKey(opts: {
...
+    /** the recipient's curve25519 key */
+    recipientCurve25519Key: string;
+    /** the recipient's ed25519 key */
+    recipientEd25519Key: string;
+    /** sender's olm account */
+    olmAccount: Olm.Account;
+    /** sender's olm session with the recipient */
+    p2pSession: Olm.Session;
+    requestId: string;
+    secret: string;
```
Comment (richvdh) [inline spec/integ/crypto/olm-utils.ts:352]: would be nice to include docs for `sender`, `requestId` and `secret`

File: spec/integ/crypto/verification.spec.ts [hunk 7]
```diff
@@ -1160,6 +1172,296 @@ describe.each(Object.entries(CRYPTO_BACKENDS))("verification (%s)", (backend: st
         );
     });
 
+    describe("Secrets are gossiped after verification", () => {
+        // we use a legacy olm session as the existing session,
```
Comment (richvdh) [inline spec/integ/crypto/verification.spec.ts:1176]:
```suggestion
        // we use a legacy olm session as the existing session.
```

https://en.wikipedia.org/wiki/Comma_splice

File: spec/integ/crypto/verification.spec.ts [hunk 8]
```diff
@@ -1221,6 +1523,64 @@ function expectSendToDeviceMessage(msgtype: string): Promise<{ messages: any }>
     });
 }
 
+/**
+ * Utility to add all needed mocks for secret requesting (to device of type `m.secret.request`).
```
Comment (richvdh) [inline spec/integ/crypto/verification.spec.ts:1527]:
```suggestion
 * Utility to add all needed mocks for secret requesting (to-device of type `m.secret.request`).
```

File: spec/integ/crypto/verification.spec.ts [hunk 9]
```diff
@@ -1221,6 +1523,64 @@ function expectSendToDeviceMessage(msgtype: string): Promise<{ messages: any }>
...
+ *
+ * The following secrets are mocked: `m.cross_signing.master`, `m.cross_signing.self_signing`,
+ * `m.cross_signing.user_signing`, `m.megolm_backup.v1`.
+ *
+ * @returns a map of secret name to promise that will resolve when the secret is requested.
+ */
+function mockUpSecretRequestAndGetPromises(): Map<string, Promise<string>> {
```
Comment (richvdh) [inline spec/integ/crypto/verification.spec.ts:1534]:
```suggestion
function mockSecretRequestAndGetPromises(): Map<string, Promise<string>> {
```

File: spec/integ/crypto/verification.spec.ts [hunk 10]
```diff
@@ -1221,6 +1523,64 @@ function expectSendToDeviceMessage(msgtype: string): Promise<{ messages: any }>
...
+    let mskRequested: (requestId: string) => void;
+    const awaitMskRequest: Promise<string> = new Promise((resolve) => {
+        mskRequested = resolve;
+    });
```
Comment (richvdh) [inline spec/integ/crypto/verification.spec.ts:1538]: this is fine, but for future reference, consider something like:

```typescript
    let mskRequestDefer = defer<string>();
    promiseMap.set("m.cross_signing.master", mskRequestDefer.promise);
    
    // ...

    if (name == "m.cross_signing.master") {
        mskRequestDefer.resolve(requestId);
    }
   
```

(`defer` from `matrix-js-sdk/src/utils.ts`). And then

File: spec/integ/crypto/verification.spec.ts [hunk 11]
```diff
@@ -1221,6 +1523,64 @@ function expectSendToDeviceMessage(msgtype: string): Promise<{ messages: any }>
     });
 }
 
+/**
+ * Utility to add all needed mocks for secret requesting (to device of type `m.secret.request`).
+ *
+ * The following secrets are mocked: `m.cross_signing.master`, `m.cross_signing.self_signing`,
+ * `m.cross_signing.user_signing`, `m.megolm_backup.v1`.
+ *
+ * @returns a map of secret name to promise that will resolve when the secret is requested.
```
Comment (richvdh) [inline spec/integ/crypto/verification.spec.ts:1532]:
```suggestion
 * @returns a map of secret name to promise that will resolve (with the id of the secret request) when the secret is requested.
```

File: spec/integ/crypto/verification.spec.ts (hunk 3 above)
Comment (richvdh) [inline spec/integ/crypto/verification.spec.ts:1273]: this looks wrong. Please let's not add half-second sleeps to all our tests!

If we're just waiting for promises to resolve, then `await jest.runAllTimersAsync();` should do fine.

Likewise everywhere below.

**Review by richvdh — CHANGES_REQUESTED** [review]

File: spec/integ/crypto/verification.spec.ts (hunk 3 above)
Comment (BillCarsonFr) [reply spec/integ/crypto/verification.spec.ts:1273]: Yes I know. But it's very temporary code as the following PR is adding signaling that we can use.
Also `jest.runAllTimersAsync()` is not working, so not sure what to do.

File: src/rust-crypto/rust-crypto.ts [hunk 12]
```diff
@@ -1357,6 +1357,50 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+
+                await this.storeSessionBackupPrivateKey(decodeBase64(value), backupCheck.backupInfo.version);
+                // XXXX at this point we should probably try to download the backup and import the keys,
+                // or at least retry for the current decryption failures?
+                // Maybe add some signaling when a new secret is received, and let clients handle it?
+                // as it's where the restore from backup APIs are
+
+                // The secret is valid and stored, clear the inbox.
+                // Important to call this after storing the secret as good hygiene.
+                await this.olmMachine.deleteSecretsFromInbox("m.megolm_backup.v1");
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts:1399]: Yes we delete all, we have one that is correct so we can discard the others. Several can be queued if several devices did sent the secret.

[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts:1399]: The API is a bit strange, and there is no method to just delete this secret it will delete all entries for that name, so potentially the valid secret that could be in queue.

[the remark this answers is not in the document]
Comment (richvdh) [reply src/rust-crypto/rust-crypto.ts:1399]: but the secret might have changed, and we might not be looking at the most recent?

[the remark this answers is not in the document]
Comment (richvdh) [reply src/rust-crypto/rust-crypto.ts:1399]: oh, that's what you mean at https://github.com/matrix-org/matrix-js-sdk/pull/3778#discussion_r1358202738

[the remark this answers is not in the document]
Comment (richvdh) [reply src/rust-crypto/rust-crypto.ts:1399]: hrm. So if we get a bad secret, it might never be deleted?

File: spec/integ/crypto/verification.spec.ts (hunk 3 above)
Comment (richvdh) [reply spec/integ/crypto/verification.spec.ts:1273]: if we're fixing it very soon I guess that's ok

File: src/rust-crypto/backup.ts [hunk 13]
```diff
@@ -106,6 +106,31 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
...
+    }
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/backup.ts:132]: Yes I tried that but the backupInfoMatchesPrivateKeyBase64 is a bit special as it's used for secret (that are received without context of version or algorithm), on the other one we have a BackupKeys that has more context. So at the end it's just factoring out an equal:
`uath_data.public_key === backupDecryptionKey.megolmV1PublicKey.publicKeyBase64`

File: src/rust-crypto/rust-crypto.ts [hunk 14]
```diff
@@ -1357,6 +1357,50 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+                    // just ignore the secret
+                    return;
+                }
+                this.logger.info(`onReceiveSecret: Received matching secret ${name}, store it.`);
+
+                await this.storeSessionBackupPrivateKey(decodeBase64(value), backupCheck.backupInfo.version);
+                // XXXX at this point we should probably try to download the backup and import the keys,
+                // or at least retry for the current decryption failures?
+                // Maybe add some signaling when a new secret is received, and let clients handle it?
+                // as it's where the restore from backup APIs are
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts]: Yes I moved it into BackupManager

File: src/rust-crypto/rust-crypto.ts [hunk 15]
```diff
@@ -1357,6 +1357,50 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+                const privateKeyMatcehs = this.backupManager.backupMatchesPrivateKey(backupCheck.backupInfo, value);
+
+                if (!privateKeyMatcehs) {
+                    this.logger.debug(`onReceiveSecret: backup decryption key does not match current backup version`);
+                    // just ignore the secret
+                    return;
+                }
+                this.logger.info(`onReceiveSecret: Received matching secret ${name}, store it.`);
+
+                await this.storeSessionBackupPrivateKey(decodeBase64(value), backupCheck.backupInfo.version);
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts]: Updated as I moved the logic in backupManager

File: src/rust-crypto/rust-crypto.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts]: Yes, I am not sure how to fix that. The annoying part is that the secret handling triggers a check backup and enable.
I guess a solution could be to just listen to the secret stream to be aware of udpates, but then manually call getSecretsFromInbox that returns an array with all secrets, then delete at the end? WDYT?
The rust API is a bit strange :/

File: spec/integ/crypto/verification.spec.ts (hunk 10 above)
Comment (BillCarsonFr) [reply spec/integ/crypto/verification.spec.ts]: Thanks for the tip, a lot better.

File: spec/integ/crypto/verification.spec.ts [hunk 16]
```diff
@@ -1266,11 +1626,15 @@ function buildRequestMessage(transactionId: string): { type: string; content: ob
 }
 
 /** build an m.key.verification.ready to-device message originating from the dummy device */
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply spec/integ/crypto/verification.spec.ts:1612]: Not sure to get this

File: src/rust-crypto/rust-crypto.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts]: I have changed the way to do it. Now listen for m.secret.send events and the poll the store with OlmMachine::store::get_secrets_from_inbox() every time one is received. Then I delete all when all as been processed

File: spec/integ/crypto/verification.spec.ts (hunk 3 above)
Comment (BillCarsonFr) [reply spec/integ/crypto/verification.spec.ts:1269]: The next PR will remove that as I am adding better signaling (crypto event)

Comment (BillCarsonFr) [reply spec/integ/crypto/verification.spec.ts:1269]: fixed as part of https://github.com/matrix-org/matrix-js-sdk/pull/3783

Comment (florianduros) [reply spec/integ/crypto/verification.spec.ts:1269]: okay fine!

**Review by florianduros — APPROVED** [review]
LGTM

File: src/rust-crypto/backup.ts [hunk 17]
```diff
@@ -106,6 +106,27 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
...
     }
 
+    /**
+     * Checks if the provided backup info matches the given private key.
+     *
+     * @param info - The backup info to check.
+     * @param backupDecryptionKey - The `BackupDecryptionKey` private key to check against.
+     * @returns `true` if the private key can decrypt the backup, `false` otherwise.
+     */
+    private backupInfoMatchesBackupDecryptionKey(
```
Comment (richvdh) [inline src/rust-crypto/backup.ts:116]: make this a standalone function, since it doesn't use `this`?

File: src/rust-crypto/backup.ts [hunk 18]
```diff
@@ -106,6 +106,27 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
...
+        info: KeyBackupInfo,
+        backupDecryptionKey: RustSdkCryptoJs.BackupDecryptionKey,
+    ): boolean {
+        if (info.algorithm !== "m.megolm_backup.v1.curve25519-aes-sha2") {
```
Comment (richvdh) [inline src/rust-crypto/backup.ts:120]: shouldn't we also check that `backupDecryptionKey.megolmV1PublicKey.algorithm` is `"m.megolm_backup.v1.curve25519-aes-sha2"`  ?

File: src/rust-crypto/backup.ts (hunk 17 above)
Comment (richvdh) [inline src/rust-crypto/backup.ts:116]: can we use this in `isKeyBackupTrusted`?  I think we just need

```
const backupMatchesSavedPrivateKey = backupKeys?.decryptionKey !== undefined && backupInfoMatchesBackupDecruptionKey(info, backupKeys.decryptionKey);
```

File: src/rust-crypto/backup.ts [hunk 19]
```diff
@@ -124,6 +145,43 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
...
+                const privateKeyMatches = this.backupInfoMatchesBackupDecryptionKey(
+                    backupCheck.backupInfo,
+                    backupDecryptionKey,
+                );
+                if (!privateKeyMatches) {
+                    logger.debug(`onReceiveSecret: backup decryption key does not match current backup version`);
+                    // just ignore the secret
+                    return false;
+                }
+                logger.info(`onReceiveSecret: Received matching secret ${secret}, store it.`);
```
Comment (richvdh) [inline src/rust-crypto/backup.ts:173]: storing the secret key in a rageshake seems unwise.

Generally this seems like a rather cryptic log line. Can you make it easier for a reader to understand what it means? Eg, what sort of secret?

File: spec/integ/crypto/verification.spec.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply spec/integ/crypto/verification.spec.ts:1612]: I'm quoting from the existing text. It doesn't appear to be accurate.

File: spec/integ/crypto/verification.spec.ts [hunk 20]
```diff
@@ -1221,6 +1519,52 @@ function expectSendToDeviceMessage(msgtype: string): Promise<{ messages: any }>
...
+    const sskRequestDefer = defer<string>();
+    const uskRequestDefer = defer<string>();
+    const backupKeyRequestDefer = defer<string>();
+
+    fetchMock.put(
+        new RegExp(`/_matrix/client/(r0|v3)/sendToDevice/m.secret.request`),
+        (url: string, opts: RequestInit): MockResponse => {
+            const messages = JSON.parse(opts.body as string).messages[TEST_USER_ID];
+            // rust crypto broadcasts to all devices, old crypto to a specific device, take the first one
+            const content = Object.values(messages)[0] as any; //messages[TEST_DEVICE_ID] ? messages[TEST_DEVICE_ID] : messages['*'];
```
Comment (richvdh) [inline spec/integ/crypto/verification.spec.ts:1541]: I don't really understand what this comment means. Is it important? Can you rephrase or delete?

**Review by richvdh — CHANGES_REQUESTED** [review]

File: src/rust-crypto/rust-crypto.ts [hunk 21]
```diff
@@ -1357,6 +1357,54 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+     *
+     * The only secret this is currently broadcasted is the `m.megolm_backup.v1`.
+     *
+     * @param name - the secret name
+     * @param base64 - the secret value base 64 encoded
+     */
+    private async handleSecretReceived(name: string, value: string): Promise<boolean> {
+        this.logger.debug(`onReceiveSecret: Received secret ${name}`);
+        if (name === "m.megolm_backup.v1") {
+            const isHandled = await this.backupManager.handleBackupSecretReceived(value);
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts:1376]: Sure was to mark the TODO as the boolean would have to be checked, but will put in next PR

File: src/rust-crypto/rust-crypto.ts [hunk 22]
```diff
@@ -1357,6 +1357,54 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+        return false;
+    }
+
+    /**
+     * Called when a new secret is received in the rust secret inbox.
+     *
+     * Will poll the secret inbox and handle the secrets received.
+     *
+     * @param name - The name of the secret received.
+     * @returns `true` if the secret has been handled and saved, `false` otherwise.
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts:1394]: udpated

File: src/rust-crypto/rust-crypto.ts [hunk 23]
```diff
@@ -1357,6 +1357,54 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+     */
+    public async checkSecrets(name: string): Promise<void> {
+        const pendingValues: string[] = await this.olmMachine.getSecretsFromInbox("m.megolm_backup.v1");
+        for (const value of pendingValues) {
+            if (await this.handleSecretReceived(name, value)) {
+                break;
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts:1400]: added

File: src/rust-crypto/rust-crypto.ts [hunk 24]
```diff
@@ -1357,6 +1357,54 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
         }
     }
 
+    /**
+     * Callback for `OlmMachine.registerReceiveSecretCallback`.
+     *
+     * Called by the rust-sdk whenever a secret is received.
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts:1363]: Updated the rest of the comment also (params)

File: src/rust-crypto/rust-crypto.ts [hunk 25]
```diff
@@ -1357,6 +1357,54 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
...
+     */
+    private async handleSecretReceived(name: string, value: string): Promise<boolean> {
+        this.logger.debug(`onReceiveSecret: Received secret ${name}`);
+        if (name === "m.megolm_backup.v1") {
+            const isHandled = await this.backupManager.handleBackupSecretReceived(value);
+
+            // XXXX at this point we should probably try to download the backup and import the keys,
+            // or at least retry for the current decryption failures?
+            // Maybe add some signaling when a new secret is received, and let clients handle it?
+            // as it's where the restore from backup APIs are exposed.
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts]: maybe let's see in the following PR adding this feature

File: src/rust-crypto/backup.ts [hunk 26]
```diff
@@ -124,6 +145,43 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
...
+     *
+     * @param secret - The secret as received from a `m.secert.send` event for secret `m.megolm_backup.v1`.
+     * @returns true if the secret is valid and has been stored, false otherwise.
+     */
+    public async handleBackupSecretReceived(secret: string): Promise<boolean> {
+        // Currently we only receive the decryption key without any key backup version, it is important to
+        // check that the secret is valid for the current version before storing it.
+        // We force a check to ensure to have the latest version. We also want to check that the backup is trusted
+        // as we don't want to store the secret if the backup is not trusted, and eventually import megolm keys later from an untrusted backup.
+        const backupCheck = await this.checkKeyBackupAndEnable(true);
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/backup.ts:160]: Not sure to get it, are you suggesting to skip the checks?. As explained in the comment we need to check for the current backup and we want it to be trusted. So yes if it's trusted it will enable it, can't hurt.

>  clients still using an old backup version

This is not possible, there is only one backup you can upload to.

File: src/rust-crypto/backup.ts [hunk 27]
```diff
@@ -124,6 +145,43 @@ export class RustBackupManager extends TypedEventEmitter<RustBackupCryptoEvents,
...
+        if (backupCheck?.backupInfo?.version && backupCheck.trustInfo.trusted) {
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/backup.ts:161]: sure

File: src/rust-crypto/backup.ts (hunk 17 above)
Comment (BillCarsonFr) [reply src/rust-crypto/backup.ts:116]: I think I got it?

Comment (BillCarsonFr) [reply src/rust-crypto/backup.ts:116]: not anymore because as per previous comment we move it to a standalone function

File: src/rust-crypto/backup.ts (hunk 19 above)
Comment (BillCarsonFr) [reply src/rust-crypto/backup.ts:173]: good catch, got lost with all the refactoring from review was the secret name initialy.

File: spec/integ/crypto/verification.spec.ts (hunk 20 above)
Comment (BillCarsonFr) [reply spec/integ/crypto/verification.spec.ts:1541]: So it's too much comment?

I mean it's important if you want to use this for legacy test. Just explains that with rust crypto the secret request is broadcasted to "*" and in legacy just sent to `TEST_DEVICE_ID`

File: spec/integ/crypto/verification.spec.ts (hunk 16 above)
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply spec/integ/crypto/verification.spec.ts]: Oh ok got it. Some clarification could help

File: src/rust-crypto/rust-crypto.ts [hunk 28]
```diff
@@ -1357,6 +1357,54 @@ export class RustCrypto extends TypedEventEmitter<RustCryptoEvents, RustCryptoEv
         }
     }
 
+    /**
+     * Callback for `OlmMachine.registerReceiveSecretCallback`.
```
[the remark this answers is not in the document]
Comment (BillCarsonFr) [reply src/rust-crypto/rust-crypto.ts]: updated

File: spec/integ/crypto/verification.spec.ts (hunk 20 above)
Comment (richvdh) [reply spec/integ/crypto/verification.spec.ts:1541]: It's specifically the `//messages[TEST_DEVICE_ID] ? messages[TEST_DEVICE_ID] : messages['*'];` that I'm complaining about. I suppose it makes sense now that I've thought about it again, but I just don't see that it's adding any value.

File: src/rust-crypto/backup.ts (hunk 26 above)
[the remark this answers is not in the document]
Comment (richvdh) [reply src/rust-crypto/backup.ts:139]: yeah, fair enough.

File: src/rust-crypto/backup.ts (hunk 17 above)
Comment (richvdh) [reply src/rust-crypto/backup.ts]: It's not a big deal, but I don't understand why that makes any difference.

Here's what I'm suggesting, in full:

```typescript
    /**
     * Determine if a key backup can be trusted.
     *
     * @param info - key backup info dict from {@link MatrixClient#getKeyBackupVersion}.
     */
    public async isKeyBackupTrusted(info: KeyBackupInfo): Promise<BackupTrustInfo> {
        const signatureVerification: SignatureVerification = await this.olmMachine.verifyBackup(info);
        const backupKeys: RustSdkCryptoJs.BackupKeys = await this.olmMachine.getBackupKeys();
        const decryptionKey = backupKeys?.decryptionKey;
        const backupMatchesSavedPrivateKey = !!decryptionKey && backupInfoMatchesBackupDecryptionKey(info, decryptionKey);
        return {
            matchesDecryptionKey: backupMatchesSavedPrivateKey,
            trusted: signatureVerification.trusted(),
        };
    }
```

does that not work?

File: spec/integ/crypto/verification.spec.ts (hunk 20 above)
Comment (richvdh) [reply spec/integ/crypto/verification.spec.ts:1541]:
```suggestion
            const content = Object.values(messages)[0] as any;
```

**Review by richvdh — APPROVED** [review]
LGTM apart from the two suggestions I have made

File: src/rust-crypto/backup.ts (hunk 17 above)
Comment (BillCarsonFr) [reply src/rust-crypto/backup.ts]: Didn't know about the `!!` trick.
Yes it should work, looks better, updated.
